contracts/utils/structs/CircularBuffer.sol

```diff
@@ -0,0 +1,114 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.20;
+
+import {Math} from "../math/Math.sol";
+import {Arrays} from "../Arrays.sol";
+import {Panic} from "../Panic.sol";
+
+/**
+ * @dev A buffer of items of fixed size. When a new item is pushed, it takes the place of the oldest one in the buffer
+ * so that at all times, only the last N elements are kept. Items cannot be removed. The entier buffer can be reset.
+ * Last N elements can be accessed using their index from the end.
+ *
+ * Complexity:
+ * - insertion (`push`): O(1)
+ * - lookup (`last`): O(1)
+ * - inclusion (`includes`): O(N) (worst case)
+ * - reset (`clear`): O(1)
+ *
+ * * The struct is called `Bytes32CircularBuffer`. Other types can be cast to and from `bytes32`. This data structure
+ * can only be used in storage, and not in memory.
+ * ```solidity
+ * CircularBuffer.Bytes32CircularBuffer buffer;
+ * ```
+ */
+library CircularBuffer {
+    /**
+     * @dev Counts the number of items that have been pushed to the buffer. The residu modulo _data.length indicates
+     * where the next value should be stored.
+     *
+     * Struct members have an underscore prefix indicating that they are "private" and should not be read or written to
+     * directly. Use the functions provided below instead. Modifying the struct manually may violate assumptions and
+     * lead to unexpected behavior.
+     *
+     * The last item is at data[(index - 1) % data.length] and the last item is at data[index % data.length]. This
+     * range can wrap around.
+     */
+    struct Bytes32CircularBuffer {
+        uint256 _count;
```

Comment:
I get that this value is needed to insert "wrap around" insertions, but I have a couple of thoughts:

- The value is uint256. Making the count uint256 is making the structure an array, it feels off that we don't have a smaller value. However, I'm not sure what we should use instead.
- The count seems to be relevant only when the structure is not full.
- Similar to Merkle Tree's root, one can track this value outside of the contract, and this seems rather an opinionated decision

I'm currently not opposed to keep it, but I'd like to reconsider it.

Reply:
2. The count is always relevant. When the structure it point to the location where the next event should be pushed.
3. No, we need it internally to push new values.

Reply:
You think it would be clearer if we named it index, or pos ?

Reply:
Nah, I think `_count` is fine. I was trying to make a case for using a smaller variable size (e.g. `uint32`). There's no point in having a `uint256` here since I don't think any CircularBuffer will ever use all the uint256 space. However, there's also no reason to pack the variable.

Let's keep it for now, but now that we have the Packing library, a crazy idea would be to make "parallel" buffers (e.g. 8 channels of uint32's)

Let's keep it as is rn